### PR TITLE
[ fix #221 ] Functor, Applicative, Monad instances for Sum

### PR DESCRIPTION
**Pull request by gallais**
I have used module parameters and named the Functors in use to make the types 
more readable. I have also thrown in `RawApplicative` instances because we might
as well have them too.

### CODE REVIEW

File: src/Data/Sum/Categorical.agda [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+open import Category.Monad
+open import Function
+
+-- To minimize the universe level of the RawFunctor, we require that elements of
+-- B are "lifted" to a copy of B at a higher universe level (a ⊔ b). See the
+-- examples for how this is done.
+
+module _ {a} (A : Set a) (b : Level) where
+
+  Sumₗ : Set (a ⊔ b) → Set (a ⊔ b)
```
Comment (JacquesCarette) [inline src/Data/Sum/Categorical.agda]: Shouldn't the first level be just `b` ?

File: src/Data/Sum/Categorical.agda [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  monadₗ = record
+    { return = inj₂
+    ; _>>=_  = λ x f → [ inj₁ , f ]′ x
+    }
+
+-- The following are the "right-handed" versions
+
+module _ (a : Level) {b} (B : Set b) where
+
+  Sumᵣ : Set (a ⊔ b) → Set (a ⊔ b)
```
Comment (JacquesCarette) [inline src/Data/Sum/Categorical.agda]: and here just `a` ?

**Comment by gallais** [comment]
From what I understand (@siddharthist correct me if I am wrong), this solution combines
the constraint that `RawMonad` expects an endofunctor and the fact that we may want the
second set to be larger than the parametric one.

**Comment by JacquesCarette** [comment]
Ah, the endofunctor constraint might indeed be at fault here. Hmm.

**Comment by siddharthist** [comment]
Yes, I believe that's right.

**Comment by JacquesCarette** [comment]
The added commits are very nice indeed.

**Comment by MatthewDaggitt** [comment]
Is everyone now happy with this?

**Pull request merged**